### DISCUSSION TITLE
How to correctly handle initMethodSelfParamTypeVar errors

### DISCUSSION THREAD

**Discussion by runfalk** [microsoft/pyright#7863] [Q&A] [answered]
Since pyright was made stricter in a recent release I have encountered `initMethodSelfParamTypeVar` for my `FrozenMap` implementation. It currently looks something like this:

```python
import collections.abc
import itertools
import typing as t

K = t.TypeVar("K")
V = t.TypeVar("V")


class FrozenMap(t.Generic[K, V], collections.abc.Mapping[K, V]):
    """An immutable dict implementation"""

    __slots__ = ("_dict", "_hash")
    _dict: dict[K, V]
    _hash: int | None

    @t.overload
    def __init__(self) -> None: ...

    @t.overload
    def __init__(self, mapping: collections.abc.Mapping[K, V]) -> None: ...

    @t.overload
    def __init__(self, it: collections.abc.Iterable[tuple[K, V]]) -> None: ...

    @t.overload
    def __init__(self: "FrozenMap[str, V]", **kwargs: V) -> None: ...

    def __init__(self, *args, **kwargs) -> None:  # type: ignore
        self._dict = dict(*args, **kwargs)
        self._hash = None

    ...
```

`def __init__(self: "FrozenMap[str, V]", **kwargs: V) -> None: ...` is no longer allowed, but unless I keep it in my `K` will be `Unknown`. Even though this now raises an error, it still resolves the correct type of `str` for the key. What is the proper way to handle this going forwards? If I add an explicit ignore, will it likely break in the future?

I'm using strict mode.

**Comment by erictraut** [comment, marked as answer]
The typing spec was recently updated to clarify that class-scoped type variables cannot be used in the `self` parameter of an `__init__` method, so pyright's error is correct here. 

You can use method-scoped type variables in this case, so that's the approach I'd recommend. Create a new type variable (say, `T`), and change your last overload signature to be:

```python
    @t.overload
    def __init__(self: "FrozenMap[str, T]", **kwargs: T) -> None: ...
```

**Reply by runfalk** [reply]
Perfect, I had misinterpreted the change as if `self` couldn't be type hinted. This does exactly what I want. Thanks!

**Reply by erictraut** [reply]
If you're interested in the spec, [here](https://typing.readthedocs.io/en/latest/spec/constructors.html#init-method) is a link.
